### PR TITLE
[x86/Linux] Funclet-based synchronization

### LINKED ISSUE #10405: [x86/Linux] JIT.Regression.CLR-x86-JIT.V1.1-M1-Beta1.b143840 test failure
The test `JIT.Regression.CLR-x86-JIT.V1.1-M1-Beta1.b143840` fails on x86/Linux.

Backtrace:
```
#0  0xf7fdb440 in __kernel_vsyscall ()
#1  0xf7ce5687 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#2  0xf7ce8ab3 in __GI_abort () at abort.c:89
#3  0xf79a6dd9 in PROCEndProcess (hProcess=0xffffff01, uExitCode=123456789, bTerminateUnconditionally=1) at /home/hanjoung/ws/dotnet/coreclr/sr
c/pal/src/thread/process.cpp:1385
#4  0xf79a6f4c in TerminateProcess (hProcess=0xffffff01, uExitCode=123456789) at /home/hanjoung/ws/dotnet/coreclr/src/pal/src/thread/process.cpp:1301
#5  0xf6d09c02 in TerminateOnAssert () at /home/hanjoung/ws/dotnet/coreclr/src/utilcode/debug.cpp:183
#6  0xf6d0aeb2 in _DbgBreakCheck (szFile=0xf79ea4c8 "/home/hanjoung/ws/dotnet/coreclr/src/vm/methodtable.cpp", iLine=9628, szExpr=0xf7[13/9211]
anityCheck()", fConstrained=0) at /home/hanjoung/ws/dotnet/coreclr/src/utilcode/debug.cpp:436
#7  0xf6d0b645 in _DbgBreakCheckNoThrow (szFile=0xf79ea4c8 "/home/hanjoung/ws/dotnet/coreclr/src/vm/methodtable.cpp", iLine=9628, szExpr=0xf79d
8abe "SanityCheck()", fConstrained=0) at /home/hanjoung/ws/dotnet/coreclr/src/utilcode/debug.cpp:548
#8  0xf6d0bce5 in DbgAssertDialog (szFile=0xf79ea4c8 "/home/hanjoung/ws/dotnet/coreclr/src/vm/methodtable.cpp", iLine=9628, szExpr=0xf79d8abe "
SanityCheck()") at /home/hanjoung/ws/dotnet/coreclr/src/utilcode/debug.cpp:735
#9  0xf6e559b2 in MethodTable::Validate (this=0xffc64e20) at /home/hanjoung/ws/dotnet/coreclr/src/vm/methodtable.cpp:9628
#10 0xf6e654f2 in Object::ValidateInner (this=0xf31b1308, bDeep=1, bVerifyNextHeader=1, bVerifySyncBlock=1) at /home/hanjoung/ws/dotnet/coreclr
/src/vm/object.cpp:1720
#11 0xf6e63ef9 in Object::Validate (this=0xf31b1308, bDeep=1, bVerifyNextHeader=1, bVerifySyncBlock=1) at /home/hanjoung/ws/dotnet/coreclr/src/
vm/object.cpp:1696
#12 0xf6e684cd in OBJECTREF::OBJECTREF (this=0xf08a2cc0, pObject=0xf31b1308) at /home/hanjoung/ws/dotnet/coreclr/src/vm/object.cpp:2691
#13 0xf6d75a66 in EECodeManager::GetInstance (this=0x8078ed0, pContext=0xf08a3370, pCodeInfo=0xf08a3188) at /home/hanjoung/ws/dotnet/coreclr/sr
c/vm/eetwain.cpp:5224
#14 0xf6eba534 in CrawlFrame::GetThisPointer (this=0xf08a2f68) at /home/hanjoung/ws/dotnet/coreclr/src/vm/stackwalk.cpp:184
#15 0xf6ec14a6 in StackFrameIterator::PreProcessingForManagedFrames (this=0xf08a2f68) at /home/hanjoung/ws/dotnet/coreclr/src/vm/stackwalk.cpp:3161
#16 0xf6ebd43a in StackFrameIterator::ProcessCurrentFrame (this=0xf08a2f68) at /home/hanjoung/ws/dotnet/coreclr/src/vm/stackwalk.cpp:2978
#17 0xf6ec0e26 in StackFrameIterator::NextRaw (this=0xf08a2f68) at /home/hanjoung/ws/dotnet/coreclr/src/vm/stackwalk.cpp:2738
#18 0xf6ebc54b in StackFrameIterator::Next (this=0xf08a2f68) at /home/hanjoung/ws/dotnet/coreclr/src/vm/stackwalk.cpp:1601
#19 0xf6ebbc53 in Thread::StackWalkFramesEx (this=0x80d8398, pRD=0xf08a3370, pCallback=0xf71d4ee0 <UpdateObjectRefInResumeContextCallback(CrawlFrame*, void*)>, pData=0xf08a32f0, flags=0, pStartFrame=0xf08a4d04) at /home/hanjoung/ws/dotnet/coreclr/src/vm/stackwalk.cpp:952
#20 0xf71d536c in ExceptionTracker::FindNonvolatileRegisterPointers (pThread=0x80d8398, uOriginalSP=4035597248, pRegDisplay=0xf08a3370, uResum$FrameFP=4035597288) at /home/hanjoung/ws/dotnet/coreclr/src/vm/exceptionhandling.cpp:374
#21 0xf71db88f in FixNonvolatileRegisters (uOriginalSP=4035597248, pThread=0x80d8398, pContextRecord=0xf08a4050, fAborting=false) at /home/hanjoung/ws/dotnet/coreclr/src/vm/exceptionhandling.cpp:1240
#22 0xf71d8e15 in ProcessCLRException (pExceptionRecord=0xeff00848, MemoryStackFp=4035597296, pContextRecord=0xf08a4050, pDispatcherContext=0xf
08a4020) at /home/hanjoung/ws/dotnet/coreclr/src/vm/exceptionhandling.cpp:1159
---Type <return> to continue, or q <return> to quit---
#23 0xf71e28d0 in UnwindManagedExceptionPass2 (ex=..., unwindStartContext=0xf08a4500) at /home/hanjoung/ws/dotnet/coreclr/src/vm/exceptionhandl
ing.cpp:4430
#24 0xf71e31e3 in UnwindManagedExceptionPass1 (ex=..., frameContext=0xf08a4888) at /home/hanjoung/ws/dotnet/coreclr/src/vm/exceptionhandling.cp
p:4592
#25 0xf71e3c77 in DispatchManagedException (ex=..., isHardwareException=false) at /home/hanjoung/ws/dotnet/coreclr/src/vm/exceptionhandling.cpp
:4699
#26 0xf705ad4c in IL_Throw (obj=0xf31b1314) at /home/hanjoung/ws/dotnet/coreclr/src/vm/jithelpers.cpp:5296
#27 0xf4f636b5 in ?? ()
#28 0xf4f63690 in ?? ()
#29 0xf4f63690 in ?? ()
#30 0xf4f63690 in ?? ()

... more frames (seems like recursive calls) ...
```

### PR DESCRIPTION
**Pull request by parjong**
This commit revises JIT to emit Monitor Exit Helper Call inside finally to fix #10405.

### CODE REVIEW

**Comment by parjong** [comment]
@jkotas @janvorli @BruceForstall Please take a look.

File: src/vm/stackwalk.cpp [hunk 1]
```diff
@@ -3149,7 +3149,10 @@ void StackFrameIterator::PreProcessingForManagedFrames(void)
 
     INDEBUG(m_crawl.pThread->DebugLogStackWalkInfo(&m_crawl, "CONSIDER", m_uFramesProcessed));
 
-#if defined(_DEBUG) && defined(_TARGET_X86_) && !defined(DACCESS_COMPILE)
+#if defined(_DEBUG) && !defined(WIN64EXCEPTIONS) && !defined(DACCESS_COMPILE)
+    //
```
Comment (BruceForstall) [inline src/vm/stackwalk.cpp]: You should fix the comment on the `#endif`

File: src/jit/codegenxarch.cpp [hunk 2]
```diff
@@ -5313,6 +5313,7 @@ void CodeGen::genCallInstruction(GenTreeCall* call)
     unsigned stackAdjustBias = 0;
 
 #if defined(_TARGET_X86_)
+#if !FEATURE_EH_FUNCLETS
     //-------------------------------------------------------------------------
```
Comment (BruceForstall) [inline src/jit/codegenxarch.cpp]: You could move the `#if defined(_TARGET_X86_)` below the `#endif // !FEATURE_EH_FUNCLETS` to avoid nested `#if`.

**Comment by BruceForstall** [comment]
Looks good. Just minor suggestions.

**Comment by BruceForstall** [comment]
@parjong looks like you need to run jit-format

**Comment by parjong** [comment]
@dotnet-bot test Ubuntu16.04 arm Cross Debug Build please

**Comment by parjong** [comment]
@dotnet-bot test Windows_NT x64 Debug Build and Test please

**Comment by parjong** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test please

**Comment by parjong** [comment]
@dotnet-bot test OSX10.12 x64 Checked Build and Test please

**Comment by parjong** [comment]
@dotnet-bot test Ubuntu16.04 arm Cross Debug Build please

**Pull request merged**